pkg/services/sqlstore/migrations/encrypt_passwords.go

```diff
@@ -0,0 +1,208 @@
+package migrations
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"github.com/go-xorm/xorm"
+	. "github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util"
+	"hash"
+	"io"
+)
+
+type AddEncryptPasswordsMigration struct {
+	MigrationBase
+}
+
+func (m *AddEncryptPasswordsMigration) Sql(dialect Dialect) string {
+	return "code migration"
+}
+
+func (m *AddEncryptPasswordsMigration) Exec(sess *xorm.Session, mg *Migrator) error {
+	var passwordRows []map[string]string
+	datasourceTypes := []string{
+		"mysql",
+		"influxdb",
+		"elasticsearch",
+		"graphite",
+		"prometheus",
+		"opentsdb",
+	}
+
+	sess.Cols("id", "password", "secure_json_data")
+	sess.Table("data_source")
+	sess.In("type", datasourceTypes)
+	sess.Where("password IS NOT NULL")
+	err := sess.Find(&passwordRows)
+	if err != nil {
+		return fmt.Errorf("password select failed: %v", err)
+	}
+
+	if err := updateRows(sess, passwordRows, "password"); err != nil {
+		return fmt.Errorf("password updates failed: %v", err)
+	}
+
+	var basicAuthRows []map[string]string
+	sess.Cols("id", "basic_auth_password", "secure_json_data")
+	sess.Table("data_source")
+	sess.In("type", datasourceTypes)
+	sess.Where("basic_auth_password IS NOT NULL")
+	err = sess.Find(&basicAuthRows)
+	if err != nil {
+		return fmt.Errorf("basic_auth_password select failed: %v", err)
+	}
+
+	if err := updateRows(sess, basicAuthRows, "basic_auth_password"); err != nil {
+		return fmt.Errorf("basic_auth_password updates failed: %v", err)
+	}
+
+	return nil
+}
+
+// addEncryptPasswordsMigration will move unencrypted passwords from password and basic_auth_password fields into
+// secure_json_data and encrypt them. This is done only for some core datasources that did not use the encrypted storage
+// until now.
+func addEncryptPasswordsMigration(mg *Migrator) {
```

Comment:
Better naming of this, the file and struct, like `DatasourcePasswordMigration` to denote it's a migration for datasources? The code can even be part of datasource_mig.go to put all migrations related to datasource in same file? Then you don't need to add the migration from migrations.go and can do it from datasource_mig.go instead.

Reply:
> The code can even be part of datasource_mig.go to put all migrations related to datasource in same file? Then you don't need to add the migration from migrations.go and can do it from datasource_mig.go instead.

Not sure about this. It seemed weird to me that migrations are added out of order and talking to @bergquist he suggested that going forward they should be ordered. Or do you mean just having new migration function in `datasource_mig.go` but calling it in correct order afterwards in `migration.go`?

Reply:
Okay wasn't aware of that suggestion. For me it should be enough that they're ordered in each file or am I missing something @bergquist? Makes it easier to find migrations related to same table.

Reply:
> For me it should be enough that they're ordered in each file

Ach did not realise that is the case actually. Wouldn't there be some cases where that could still break? Like merging some fields from 2 different tables or some data migration based on some values from different table? I always worked with migration system where each migrations was somehow uniquely marked and ordered and append only so that is why this seems strange to me and wonder if there could be some issues from that.

Reply:
As the migration was removed, not an issue right now. We can probably have a discussion in backend weekly at some point whether to change the migrations or not.